### PR TITLE
Improvements to the Describe > Two\Three Variable > Summaries dialog

### PR DESCRIPTION
**Pull request by Ivanluv**
This replaces PR #7343.
 @lloyddewit  I have addressed the comments you had from PR #7343 in this changes.

### CODE REVIEW

**Comment by rdstern** [comment]
@Ivanluv one small point is that when there are 2 factors there is a black sort of flash as the dialogue changes size.
More important is that this doesn't seem to work!  With the usual survey data and the factors I get as follows:
image

**Comment by Ivanluv** [comment]
@rdstern  could you have a look again?

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 1]
```diff
@@ -1378,103 +1377,112 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
       }
     } else {
       outer_margins <- NULL
     }
     if ("summary" %in% margins || ("outer" %in% margins && length(factors) == 0)) {
       summary_margins <- NULL
-      if ("outer" %in% margins) {
-        power_sets_summary <- power_sets
+      if (is.null(columns_to_summarise)){
+          power_sets_summary <- power_sets[-(length(power_sets))]
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1413]: indentation?

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 2]
```diff
@@ -1378,103 +1377,112 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
       } else {
-        power_sets_summary <- power_sets[(c(length(power_sets)))]
+          if ("outer" %in% margins) {
+              power_sets_summary <- power_sets
+          } else {
+              power_sets_summary <- power_sets[(c(length(power_sets)))]
+          }
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1419]: indentation?

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 3]
```diff
@@ -1378,103 +1377,112 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
+
+      if (is.null(columns_to_summarise)){
+        for (facts in power_sets_summary) {
+          if (length(facts) == 0) facts <- c()
+          summary_margins_df <- data_book$get_data_frame(data_name = data_name) %>%
+            dplyr::select(c(factors))
+          data_book$import_data(data_tables = list(summary_margins_df = summary_margins_df))
+          summary_margins[[length(summary_margins) + 1]] <- data_book$calculate_summary(data_name = "summary_margins_df", columns_to_summarise = NULL, summaries = summaries, factors = facts, store_results = FALSE, drop = drop, na.rm = na.rm, return_output = TRUE, weights = weights, result_names = result_names, percentage_type = percentage_type, perc_total_columns = perc_total_columns, perc_total_factors = perc_total_factors, perc_total_filter = perc_total_filter, perc_decimal = perc_decimal, margin_name = margin_name, additional_filter = additional_filter, perc_return_all = FALSE, ...)
+          data_book$delete_dataframes(data_names = "summary_margins_df")
+        }
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1430]: @lilyclements 
Most of this code is duplicated below.
If we move the `if (is.null(columns_to_summarise))` check inside the `for loop`, then can we remove the duplication?

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 4]
```diff
@@ -1378,103 +1377,112 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
           # TODO: if the colname is the same as a factor, then do nothing
           colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
-        }
-        summary_margins <- summary_margins %>% dplyr::mutate(dplyr::across(where(is.numeric), round, signif_fig))
-        summary_margins <- summary_margins %>%
-          tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value", values_transform = list(value = as.character))
+              }
+              summary_margins <- summary_margins %>%
+              tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
+          }
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1461]: indentation?

File: instat/dlgDescribeTwoVariable.vb [hunk 5]
```diff
@@ -151,101 +357,249 @@ Public Class dlgDescribeTwoVariable
...
+                End If
+                If (lstFirstItemTypes.Count > 0) Then
+                    strFirstVariablesType = lstFirstItemTypes(0)
+                Else
+                    strFirstVariablesType = ""
+                    lblFirstType.Text = "________"
+                    lblFirstType.ForeColor = SystemColors.ControlText
+                End If
+                lblFirstType.Text = strFirstVariablesType
+                lblFirstType.ForeColor = SystemColors.Highlight
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:394]: will this overwrite the assignments on lines 390 and 391?

File: instat/dlgDescribeTwoVariable.vb [hunk 6]
```diff
@@ -151,101 +357,249 @@ Public Class dlgDescribeTwoVariable
...
+            Dim iPosition As Integer = 0
+            clsCombineMultipleColumnsFunction.ClearParameters()
+            clsCombineFactorsFunction.ClearParameters()
+
+            For Each strColumn In ucrReceiverFirstVars.GetVariableNamesList()
+                clsCombineMultipleColumnsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                If ucrReceiverSecondVar.GetVariableNames <> strColumn Then
+                    clsCombineFactorsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                End If
+                iPosition = iPosition + 1
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:465]:
```suggestion
                iPosition += 1
```

File: instat/dlgDescribeTwoVariable.vb [hunk 7]
```diff
@@ -151,101 +357,249 @@ Public Class dlgDescribeTwoVariable
...
+    Private Sub Controls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverFirstVars.ControlContentsChanged, ucrReceiverSecondVar.ControlContentsChanged, ucrPnlDescribe.ControlContentsChanged
+        If Not ucrReceiverFirstVars.IsEmpty AndAlso (ucrChangedControl Is ucrReceiverFirstVars OrElse ucrChangedControl Is ucrReceiverSecondVar) Then
+            Dim iPosition As Integer = 0
+            clsCombineMultipleColumnsFunction.ClearParameters()
+            clsCombineFactorsFunction.ClearParameters()
+
+            For Each strColumn In ucrReceiverFirstVars.GetVariableNamesList()
+                clsCombineMultipleColumnsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                If ucrReceiverSecondVar.GetVariableNames <> strColumn Then
+                    clsCombineFactorsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:463]: If we use `iPosition` here, then the positions may not be consecutive.
Does this matter? Or is it cleaner to have a seprate variable for the `clsCombineFactorsFunction` parameters?

File: instat/dlgDescribeTwoVariable.vb [hunk 8]
```diff
@@ -151,101 +357,249 @@ Public Class dlgDescribeTwoVariable
...
+                End If
+                iPosition = iPosition + 1
+            Next
+            clsCombineFactorsFunction.AddParameter(ucrReceiverSecondVar.GetVariableNames, ucrReceiverSecondVar.GetVariableNames, bIncludeArgumentName:=False, iPosition:=iPosition + 1)
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:467]: do we need to increment `iPosition`?

File: instat/dlgDescribeTwoVariable.vb [hunk 9]
```diff
@@ -151,101 +357,249 @@ Public Class dlgDescribeTwoVariable
...
         End If
     End Sub
+
+    Private Sub EnableDisableFrequencyControls()
+        If rdoCustomize.Checked Then
+            If strFirstVariablesType = "categorical" AndAlso strSecondVariableType = "categorical" Then
+                grpDisplay.Visible = True
+                grpFrequency.Visible = True
+                ucrChkDisplayMargins.Visible = True
+                ucrInputMarginName.Visible = ucrChkDisplayMargins.Checked
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:532]: more readable?
```suggestion
                ucrInputMarginName.Visible = True
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Ivanluv Thanks for the new PR, a couple of comments still open from PR #7343, and a couple of new ones

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R (hunk 1 above)
Comment (Ivanluv) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R:1413]: Done

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R (hunk 2 above)
Comment (Ivanluv) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R]: Done

File: instat/dlgDescribeTwoVariable.vb (hunk 5 above)
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb:392]: Yes, I have made the correction

File: instat/dlgDescribeTwoVariable.vb (hunk 6 above)
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb]: Done

File: instat/dlgDescribeTwoVariable.vb (hunk 7 above)
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb:461]: The positions don't matter

File: instat/dlgDescribeTwoVariable.vb (hunk 8 above)
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb]: No, thanks for the reminder. That's now fixed

**Review by rdstern — APPROVED** [review]
This is now working and it would be great if it can be included.
The funny jump in the dialogue, as it enlarges is still there.  But if the code is acceptable to @lloyddewit or others, then I would like to merge, even if another issue is added.

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 10]
```diff
@@ -1378,103 +1377,114 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
+      if (is.null(columns_to_summarise)){
+          power_sets_summary <- power_sets[-(length(power_sets))]
+      } 
+      else {
+          if ("outer" %in% margins) {
+              power_sets_summary <- power_sets
+          }
+          else {
+              power_sets_summary <- power_sets[(c(length(power_sets)))]
+          }
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1421]:
```suggestion
        power_sets_summary <- power_sets[-(length(power_sets))]
      } 
      else {
        if ("outer" %in% margins) {
          power_sets_summary <- power_sets
        }
        else {
          power_sets_summary <- power_sets[(c(length(power_sets)))]
        }
```

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 11]
```diff
@@ -1378,103 +1377,114 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
+              summary_margins <- data.frame(summary_margins, `summary-variable` = "count", factors = NA)
+              names(summary_margins) <- c("value", "summary-variable", factors)
+          }else {
+              for (col in 1:ncol(summary_margins)) {
+                # TODO: if the colname is the same as a factor, then do nothing
+                colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
+              }
+              summary_margins <- summary_margins %>%
+              tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
+          }
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/Backend_Components/summary_functions.R:1463]:
```suggestion
        if (length(summary_margins) == 1) {
          summary_margins <- data.frame(summary_margins, `summary-variable` = "count", factors = NA)
          names(summary_margins) <- c("value", "summary-variable", factors)
        }else {
          for (col in 1:ncol(summary_margins)) {
            # TODO: if the colname is the same as a factor, then do nothing
            colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
          }
          summary_margins <- summary_margins %>%
          tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
        }
```

File: instat/dlgDescribeTwoVariable.vb (hunk 9 above)
Comment (lloyddewit) [reply instat/dlgDescribeTwoVariable.vb:531]: yes, my mistake, sorry

**Comment by lloyddewit** [comment]
@lilyclements or @Ivanluv If you can resolve the last open comment above, then I can approve
Thanks

**Comment by Ivanluv** [comment]
@lloyddewit @lilyclements is the best person to look on that open comment. I suggest we have this merged then I can write an issue on that

**Comment by lloyddewit** [comment]
@lilyclements Please could you resolve the last remaining comment in a new PR (when you're back from leave, no hurry)?

@rdstern Normally I'd ask you to approve again but the changes since your last approval should not affect functionality, and it will be good to include this PR in this week's release build. So I'll go ahead and merge.

**Pull request merged**